File: src/glados/routes/runs.py
"""
Runs Routes

REST endpoints for run management.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status

from src.glados.exceptions import RunNotFoundError
from src.glados.schemas import RunCreate, RunListResponse, RunResponse, RunStatus
from src.glados.services.run_manager import Run, RunManager

router = APIRouter(prefix="/api/v1/runs", tags=["runs"])

# Shared RunManager instance for MVP-2 (will be injected via DI in production)
_run_manager: RunManager | None = None


def get_run_manager() -> RunManager:
    """Get or create RunManager instance."""
    global _run_manager
    if _run_manager is None:
        _run_manager = RunManager()
    return _run_manager


def reset_run_manager() -> None:
    """Reset RunManager (for testing)."""
    global _run_manager
    _run_manager = None


def _run_to_response(run: Run) -> RunResponse:
    """Convert internal Run to RunResponse."""
    return RunResponse(
        id=run.id,
        strategy_id=run.strategy_id,
        mode=run.mode,
        status=run.status,
        symbols=run.symbols,
        timeframe=run.timeframe,
        config=run.config,
        created_at=run.created_at,
        started_at=run.started_at,
        stopped_at=run.stopped_at,
    )


@router.get("", response_model=RunListResponse)
async def list_runs(
    run_manager: RunManager = Depends(get_run_manager),
) -> RunListResponse:
    """
    List all runs.
    
    MVP-2: No pagination, returns all runs.
    """
    runs, total = await run_manager.list()
    return RunListResponse(
        items=[_run_to_response(r) for r in runs],
        total=total,
    )


@router.post("", response_model=RunResponse, status_code=status.HTTP_201_CREATED)
async def create_run(
    request: RunCreate,
    run_manager: RunManager = Depends(get_run_manager),
) -> RunResponse:
    """Create a new trading run."""
    run = await run_manager.create(request)
    return _run_to_response(run)


@router.get("/{run_id}", response_model=RunResponse)
async def get_run(
    run_id: str,
    run_manager: RunManager = Depends(get_run_manager),
) -> RunResponse:
    """Get run by ID."""
    run = await run_manager.get(run_id)
    if run is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run not found: {run_id}",
        )
    return _run_to_response(run)


@router.post("/{run_id}/stop", response_model=RunResponse)
async def stop_run(
    run_id: str,
    run_manager: RunManager = Depends(get_run_manager),
) -> RunResponse:
    """Stop a running run."""
    try:
        run = await run_manager.stop(run_id)
        return _run_to_response(run)
    except RunNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Run not found: {run_id}",
        )

File: haro/src/api/runs.ts
/**
 * Runs API
 *
 * API functions for managing trading runs.
 */

import { get, post } from "./client";
import type { Run, RunCreate, RunListResponse } from "./types";

/**
 * Fetch paginated list of runs
 */
export async function fetchRuns(params?: {
  page?: number;
  page_size?: number;
  status?: string;
}): Promise<RunListResponse> {
  const queryParams: Record<string, string> = {};
  if (params?.page) queryParams.page = String(params.page);
  if (params?.page_size) queryParams.page_size = String(params.page_size);
  if (params?.status) queryParams.status = params.status;

  return get<RunListResponse>("/runs", queryParams);
}

/**
 * Fetch a single run by ID
 */
export async function fetchRun(runId: string): Promise<Run> {
  return get<Run>(`/runs/${runId}`);
}

/**
 * Create a new run
 */
export async function createRun(data: RunCreate): Promise<Run> {
  return post<Run>("/runs", data);
}

/**
 * Start a run
 */
export async function startRun(runId: string): Promise<Run> {
  return post<Run>(`/runs/${runId}/start`);
}

/**
 * Stop a run
 */
export async function stopRun(runId: string): Promise<Run> {
  return post<Run>(`/runs/${runId}/stop`);
}
